Add limited ORC reader support

usurai opened the pull request:
This PR is an attempt to adding ORC reader support based on `DwrfReader`. For more context, please refer to #1747.

This PR introduces the following changes:
- Import `.proto` file of ORC from [apache/orc](https://github.com/apache/orc)
- Distinguish ORC and DWRF in `FileFormat` and make `DwrfReader` aware of the file format it's processing
- Adding the new `PostScript` class that is shared by of DWRF and ORC*
- Don't try to read cache when parsing ORC's `PostScript`
- For ORC's int DIRECT decoding, use RLE insterad of `DirectDecoder`

Currently I use ad hoc method to test the change, that is, try reading and printing the content of [apache/orc's example files](https://github.com/apache/orc/tree/main/examples) (excluding corrupted and encrypted) using modified `DwrfReader` and comparing the output with the output of [apache/orc's orc-contents](https://orc.apache.org/docs/cpp-tools.html#orc-contents), with above changes, I've got some of tests emitting correct output(24 out of 39). For the incorrect / crashed test cases, I think they are due to missing type support (DECIMAL, DATE etc) or feature support (bloom filter).

<details>
<summary>Detailed test cases</summary>

<google-sheets-html-origin><meta name="sheets-banding" data-sheets-range="A1:Z1000" data-sheets-banding-properties="{&quot;1&quot;:{&quot;1&quot;:{&quot;4&quot;:{&quot;1&quot;:2,&quot;2&quot;:12434877}},&quot;2&quot;:{&quot;4&quot;:{&quot;1&quot;:2,&quot;2&quot;:16777215}},&quot;3&quot;:1,&quot;4&quot;:{&quot;4&quot;:{&quot;1&quot;:2,&quot;2&quot;:15987699}},&quot;5&quot;:1}}">

Test | Correct | Issue | Compression | Type
-- | -- | -- | -- | --
TestOrcFile.testSeek | Y |   | zlib | struct<boolean1:boolean,byte1:tinyint,short1:smallint,int1:int,long1:bigint,float1:float,double1:double,bytes1:binary,string1:string,middle:struct<list:array<struct<int1:int,string1:string>>>,list:array<struct<int1:int,string1:string>>,map:map<string,struct<int1:int,string1:string>>>
TestOrcFile.test1 | Y |   | zlib | struct<boolean1:boolean,byte1:tinyint,short1:smallint,int1:int,long1:bigint,float1:float,double1:double,bytes1:binary,string1:string,middle:struct<list:array<struct<int1:int,string1:string>>>,list:array<struct<int1:int,string1:string>>,map:map<string,struct<int1:int,string1:string>>>
TestOrcFile.testStringAndBinaryStatistics | Y |   | zlib | struct<bytes1:binary,string1:string>
nulls-at-end-snappy | Y |   | snappy | struct<_col0:tinyint,_col1:smallint,_col2:int,_col3:bigint,_col4:float,_col5:double,_col6:boolean>
TestStringDictionary.testRowIndex | Y |   | zlib | struct<str:string>
TestOrcFile.testMemoryManagementV11 | Y |   | none | struct<int1:int,string1:string>
TestOrcFile.columnProjection | Y |   | none | struct<int1:int,string1:string>
TestOrcFile.emptyFile | Y |   | none | none
TestOrcFile.metaData | Y |   | none | struct<boolean1:boolean,byte1:tinyint,short1:smallint,int1:int,long1:bigint,float1:float,double1:double,bytes1:binary,string1:string,middle:struct<list:array<struct<int1:int,string1:string>>>,list:array<struct<int1:int,string1:string>>,map:map<string,struct<int1:int,string1:string>>>
demo-12-zlib | Y |   | zlib | struct<_col0:int,_col1:string,_col2:string,_col3:string,_col4:int,_col5:string,_col6:int,_col7:int,_col8:int>
TestOrcFile.testWithoutIndex | Y |   | snappy | struct<int1:int,string1:string>
zero | Y |   | none | none
TestOrcFile.testPredicatePushdown | Y |   | none | struct<int1:int,string1:string>
complextypes_iceberg | Y |   | zlib | struct<id:bigint,int_array:array<int>,int_array_array:array<array<int>>,int_map:map<string,int>,int_map_array:array<map<string,int>>,nested_struct:struct<a:int,b:array<int>,c:struct<d:array<array<struct<e:int,f:string>>>>,g:map<string,struct<h:struct<i:array<double>>>>>>
TestOrcFile.testStripeLevelStats | Y |   | zlib | struct<int1:int,string1:string>
TestOrcFile.testMemoryManagementV12 | Y |   | none | struct<int1:int,string1:string>
TestVectorOrcFile.testLzo | Y |   | lzo | struct<x:bigint,y:int,z:bigint>
TestVectorOrcFile.testZstd.0.12 | Y |   | zstd | struct<x:bigint,y:int,z:bigint>
demo-11-zlib | Y |   | zlib | struct<_col0:int,_col1:string,_col2:string,_col3:string,_col4:int,_col5:string,_col6:int,_col7:int,_col8:int>
TestVectorOrcFile.testLz4 | Y |   | lz4 | struct<x:bigint,y:int,z:bigint>
TestOrcFile.testSnappy | Y |   | snappy | struct<int1:int,string1:string>
version1999 | Y |   | none | struct<>
orc_no_format | Y |   | zlib | struct<_col0:array<string>,_col1:map<int,string>,_col2:struct<name:string,score:int>>
demo-11-none | Y |   | none | struct<_col0:int,_col1:string,_col2:string,_col3:string,_col4:int,_col5:string,_col6:int,_col7:int,_col8:int>
TestOrcFile.testTimestamp | N | TIMESTAMP | zlib | timestamp
orc_split_elim_new | N | DECIMAL | zlib | struct<userid:bigint,string1:string,subtype:double,decimal1:decimal(16,6),ts:timestamp>
decimal64_v2 | N | DECIMAL | zlib | struct<a:bigint,b:decimal(12,0),c:decimal(20,2),d:decimal(12,2),e:decimal(2,2)>
orc-file-11-format | N | DECIMAL | none | struct<boolean1:boolean,byte1:tinyint,short1:smallint,int1:int,long1:bigint,float1:float,double1:double,bytes1:binary,string1:string,middle:struct<list:array<struct<int1:int,string1:string>>>,list:array<struct<int1:int,string1:string>>,map:map<string,struct<int1:int,string1:string>>,ts:timestamp,decimal1:decimal(0,0)>
TestOrcFile.testUnionAndTimestamp | N | TIMESTAMP, UNION | none | struct<time:timestamp,union:uniontype<int,string>,decimal:decimal(38,18)>
orc_split_elim | N | DECIMAL | none | struct<userid:bigint,string1:string,subtype:double,decimal1:decimal(0,0),ts:timestamp>
TestOrcFile.testDate1900 | N | TIMESTAMP, DATE | zlib | struct<time:timestamp,date:date>
bad_bloom_filter_1.6.11 | N | bloom filter | zlib | struct<id:bigint,name:string>
TestOrcFile.testDate2038 | N | TIMESTAMP, DATE | zlib | struct<time:timestamp,date:date>
over1k_bloom | N | DECIMAL, bloom filter | zlib | struct<_col0:tinyint,_col1:smallint,_col2:int,_col3:bigint,_col4:float,_col5:double,_col6:boolean,_col7:string,_col8:timestamp,_col9:decimal(4,2),_col10:binary>
orc_split_elim_cpp | N | TIMESTAMP, DECIMAL | zlib | struct<userid:bigint,string1:string,subtype:double,decimal1:decimal(16,6),ts:timestamp>
decimal64_v2_cplusplus | N | DECIMAL | zlib | struct<a:bigint,b:decimal(10,2),c:decimal(2,2),d:decimal(2,2),e:decimal(2,2)>
bad_bloom_filter_1.6.0 | N | bloom filter | zlib | struct<id:bigint,name:string>
decimal | N | DECIMAL | none | struct<_col0:decimal(10,5)>
orc_index_int_string | N | VARCHAR | zlib | struct<_col0:int,_col1:varchar(4)>

</details>

To print an ORC file
```
_build/release/velox/examples/velox_example_scan_orc {path_to_orc_file}
```

A more important point I wanna discuss is the modeling of `proto`:
> Adding the new `PostScript` class that is shared by of DWRF and ORC

As we introduce ORC's `.proto` file, a mechanism is needed to switching between two `.proto` files flexibly. My current solution is to create a class that mimic the `proto` class' intefaces (I only implemented `PostScript` in this PR), which contains attributes of both DWRF and ORC.

zzhao0 commented on velox/dwio/dwrf/common/Common.h:
```diff
@@ -257,4 +261,54 @@ class StripeInformation {
+  virtual bool has_footerlength() const {
+    return true;
+  }
+  virtual CompressionKind compression() const {
```
I don't feel they need to be virtual methods

zzhao0 commented on velox/dwio/dwrf/common/DwrfMetadata.h:
```diff
@@ -0,0 +1,73 @@
+        cacheMode_{static_cast<StripeCacheMode>(cacheMode)},
+        cacheSize_{cacheSize} {}
+
+  explicit DwrfPostScript(std::unique_ptr<proto::PostScript>& ps)
```
Change to const proto::PostScript& ?

zzhao0 commented on velox/dwio/dwrf/reader/ReaderBase.h:
```diff
@@ -19,14 +19,17 @@
 namespace facebook::velox::dwrf {
 
-constexpr uint64_t DEFAULT_COMPRESSION_BLOCK_SIZE = 256 * 1024;
+using dwio::common::FileFormat;
```
let's avoid this in .h file

zzhao0 commented on velox/dwio/dwrf/common/DwrfMetadata.h:
```diff
@@ -0,0 +1,73 @@
+
+namespace facebook::velox::dwrf {
+
+class DwrfPostScript : public PostScript {
```
I don't feel we actually need inheritance here. Particularly for postScript, difference in two format variants are pretty small, we can just roll all properties into the same PostScript class

usurai replied:
Sure, my intention was to demonstrate my proposal to solve two protobuf models problem. For later classes (like Footer), the divergence will be bigger and I think we will eventually need some methods to deal with it.

zzhao0 commented on velox/dwio/dwrf/common/OrcMetadata.h:
```diff
@@ -0,0 +1,66 @@
+            ps->stripestatisticslength()) {}
+
+ private:
+  uint64_t metadataLength_;
```
just add these members to PostScript

zzhao0 commented on velox/dwio/dwrf/reader/DwrfReaderShared.h:
```diff
@@ -43,6 +43,7 @@ class DwrfRowReaderShared : public StrideIndexProvider,
   uint64_t strideIndex_;
   std::shared_ptr<StripeDictionaryCache> stripeDictionaryCache_;
   dwio::common::RowReaderOptions options_;
+  FileFormat fileFormat_;
```
Do we really need this member variable?

usurai replied:
I thought it would be useful when RowReader needs to deal with the difference between DWRF/ORC, but my change has not touch that part. Removed.

zzhao0 commented on velox/dwio/dwrf/reader/StripeStream.h:
```diff
@@ -91,6 +91,10 @@ class StripeStreams {
  public:
   virtual ~StripeStreams() = default;
 
+  virtual FileFormat getFormat() const {
```
why not leaving it as abstract?

usurai replied:
Yes it should be abstract.
Moved this to `StripeStreamsBase`.

zzhao0 reviewed (commented):
Let's add one simple ORC file so we have at least some test coverage.

usurai commented on velox/dwio/dwrf/common/Common.h:
```diff
@@ -257,4 +261,54 @@ class StripeInformation {
 
+class PostScript {
+ public:
+  PostScript()
```
My intent was to set default value to members, change to inline.

usurai commented on velox/dwio/dwrf/common/Common.h:
```diff
@@ -257,4 +261,54 @@ class StripeInformation {
+        writerVersion_{static_cast<WriterVersion>(writerVersion)} {}
+
+  virtual ~PostScript() = default;
+  virtual uint64_t footerlength() const {
```
Yes, this mimics protobuf style and can keep the caller code unchanged.
Have changed as suggested.

usurai commented:
Hi Zhenyuan @zzhao0, thanks for reviewing! I have made changes as suggested:
1. Add [TestStringDictionary.testRowIndex.orc](https://github.com/apache/orc/blob/main/examples/TestStringDictionary.testRowIndex.orc) from apache-orc repo and corresponding test.
2. Discard inheritance in `PostScript`.
3. Use camel style for methods of `PostScript`.
4. Misc language fixes.

Please let me know if you have any concern. Thanks!

zzhao0 commented on velox/dwio/dwrf/common/Common.h:
```diff
@@ -271,4 +275,114 @@ class StripeInformation {
   virtual uint64_t getNumberOfRows() const = 0;
 };
 
+static CompressionKind convertCompressionKind(
```
can we make this static mapping using switch statement instead? is this intended to be used outside of this scope? if not let's put it in anonymous namespace, and move the definition to cpp file

usurai replied:
Done, this indeed should be placed in unnamed namespace.

zzhao0 commented on velox/dwio/dwrf/common/Common.h:
```diff
@@ -271,4 +275,114 @@ class StripeInformation {
+            static_cast<CompressionKind>(ps.compression()),
+            ps.compressionblocksize(),
+            ps.writerversion()) {
+    cacheMode_ = static_cast<StripeCacheMode>(ps.cachemode());
```
why not make them also part of the initializater list?

usurai replied:
My intention was to make use of delegate ctor. Have changed to use initializer list.

zzhao0 commented on velox/dwio/dwrf/reader/ReaderBase.cpp:
```diff
@@ -118,11 +129,21 @@ ReaderBase::ReaderBase(
 
-  postScript_ = ProtoUtils::readProto<proto::PostScript>(
-      input_->read(fileLength_ - psLength_ - 1, psLength_, LogType::FOOTER));
+  std::variant<DwrfPostScriptPtr, OrcPostScriptPtr> postScriptProto;
```
it's the PostScript class now, there is no need of this

zzhao0 commented on velox/dwio/dwrf/reader/ReaderBase.cpp:
```diff
@@ -158,12 +186,15 @@ ReaderBase::ReaderBase(
         ->readFully(cacheBuffer->data(), cacheSize);
     cache_ = std::make_unique<StripeMetadataCache>(
-        *postScript_, *footer_, std::move(cacheBuffer));
+        *std::get<DwrfPostScriptPtr>(postScriptProto),
```
Let's change it to take PostScript class instead of proto

zzhao0 commented on velox/dwio/dwrf/reader/ReaderBase.h:
```diff
@@ -247,6 +255,7 @@ class ReaderBase {
   std::unique_ptr<encryption::DecryptionHandler> handler_;
   std::shared_ptr<dwio::common::BufferedInputFactory> bufferedInputFactory_;
 
+  dwio::common::FileFormat fileFormat_;
```
nit: since we already have a flat in PostScript class, why not push this to PostScript, and here we can just return whatever is captured in PostScript?

usurai replied:
Done.

zzhao0 commented on velox/dwio/dwrf/common/Common.h:
```diff
@@ -271,4 +275,114 @@ class StripeInformation {
+
+ private:
+  // General attributes
+  bool isDwrf_ = true;
```
cross commenting here, we can make this fileFormat_

zzhao0 commented on velox/dwio/dwrf/reader/ReaderBase.cpp:
```diff
@@ -23,12 +23,15 @@
 using dwio::common::encryption::DecrypterFactory;
 using encryption::DecryptionHandler;
 using memory::MemoryPool;
+using DwrfPostScriptPtr = std::unique_ptr<proto::PostScript>;
```
We should be able to remove this

zzhao0 commented:
@usurai please remember to rebase (since we have some refactory work going on) and fix failures in mac build

usurai commented:
> @usurai please remember to rebase (since we have some refactory work going on) and fix failures in mac build

@zzhao0 Thanks again for the detailed review! I've addressed the comments. To fix the macOS build failure, I added a new wrapper for orc's proto file. Thanks!
